### ISSUE TITLE
[Bug]: Property name "$id" leads to TypeError "unhashable type: 'dict'"

### ISSUE THREAD

**Issue by GeraldIr** [python-openapi/openapi-core#608]
### Actual Behavior

When a property named "$id" appears inside of an openAPI specification the openapi-core library fails whenever this property is part of a response validation (so when the field COULD appear in a response, it does not have to actually be in a response or be a required field). 

```
"json_schema": {
                "type": "object",
                "title": "JSON Schema",
                "description": "A JSON Schema compliant to [JSON Schema draft-07](https://json-schema.org/draft-07/json-schema-validation.html) or later.\n\nJSON Schemas SHOULD always be dereferenced (i.e. all `$refs` should be resolved).\nThis allows clients to consume the schemas much better.\nClients are not expected to support dereferencing `$refs`.\n\nNote: The specified schema in the OpenAPI document is only a common subset of JSON Schema.\nAdditional keywords from the JSON Schema specification MAY be used.",
                "properties": {
                    "$schema": {
                        "description": "The JSON Schema version. If not given in the context of openEO,\ndefaults to `draft-07`.\n\nYou may need to add the default value for `$schema` property explicitly to the JSON Schema\nobject before passing it to a JSON Schema validator.",
                        "type": "string",
                        "format": "uri",
                        "default": "http://json-schema.org/draft-07/schema#"
                    },
                    "$id": {
                        "description": "ID of your JSON Schema.",
                        "type": "string",
                        "format": "uri"
                    },
                    "type": {
                        "description": "The allowed basic data type(s) for a value.\n\nIf this property is not present, all data types are allowed.",
                        "oneOf": [
                            {
                                "$ref": "#/components/schemas/json_schema_type"
                            },
                            {
                                "type": "array",
                                "minItems": 1,
                                "uniqueItems": true,
                                "items": {
                                    "$ref": "#/components/schemas/json_schema_type"
                                }
                            }
                        ]
                    },
                    "pattern": {
                        "type": "string",
                        "format": "regex",
                        "description": "The regular expression a string value must match against."
                    },
                    "enum": {
                        "type": "array",
                        "items": {},
                        "description": "An exclusive list of allowed values."
                    },
                    "minimum": {
                        "type": "number",
                        "description": "The minimum value (inclusive) allowed for a numerical value."
                    },
                    "maximum": {
                        "type": "number",
                        "description": "The maximum value (inclusive) allowed for a numerical value."
                    },
                    "minItems": {
                        "type": "number",
                        "minimum": 0,
                        "default": 0,
                        "description": "The minimum number of items required in an array."
                    },
                    "maxItems": {
                        "type": "number",
                        "minimum": 0,
                        "description": "The maximum number of items required in an array."
                    },
                    "items": {
                        "description": "Specifies schemas for the items in an array.",
                        "anyOf": [
                            {
                                "type": "array",
                                "minItems": 1,
                                "items": {
                                    "$ref": "#/components/schemas/json_schema"
                                }
                            },
                            {
                                "$ref": "#/components/schemas/json_schema"
                            }
                        ]
                    }
                },
                "additionalProperties": {
                    "description": "You can add any other property supported by the JSON Schema version that is given through the property `$schema`, so either [draft-07](https://json-schema.org/draft-07/json-schema-validation.html) or any later version."
                }
            },
```
(Taken from https://api.openeo.org/)

### Expected Behavior

There should be no issue with properties named "$id", it should just work like any other name.

### Steps to Reproduce

Add a property named "$id" to an internal schema object and validate a response that uses that schema.

### OpenAPI Core Version

0.17.2

### OpenAPI Core Integration

Requests

### Affected Area(s)

Validation

### References

https://github.com/OAI/OpenAPI-Specification/discussions/3302

### Would you like to implement a fix?

None

**Comment by GeraldIr** [comment]
For more reference here is the Stacktrace:

Exception has occurred: TypeError       (note: full exception trace is shown but execution is paused at: test_MOCK_processes)
unhashable type: 'dict'
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/validators.py", line 878, in _find_in_subschemas
    target_uri = self._urljoin_cache(
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/validators.py", line 894, in resolve
    match = self._find_in_subschemas(url)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/_validators.py", line 294, in ref
    scope, resolved = validator.resolver.resolve(ref)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/_validators.py", line 362, in allOf
    yield from validator.descend(instance, subschema, schema_path=index)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_schema_validator/_validators.py", line 98, in allOf
    yield from _allOf(validator, allOf, instance, schema)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_schema_validator/_validators.py", line 150, in items
    yield from validator.descend(item, items, path=index)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/_validators.py", line 332, in properties
    yield from validator.descend(
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_core/validation/schemas/validators.py", line 38, in validate
    errors = tuple(errors_iter)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_core/validation/validators.py", line 106, in _validate_schema
    validator.validate(value)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_core/validation/validators.py", line 130, in _get_content_value
    self._validate_schema(schema, casted)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_core/validation/response/validators.py", line 119, in _get_data
    return self._get_content_value(raw_data, mimetype, content)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_core/validation/decorators.py", line 31, in wrapper
    return f(*args, **kwds)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_core/validation/response/validators.py", line 54, in _iter_errors
    self._get_data(data, mimetype, operation_response)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_core/validation/response/validators.py", line 192, in validate
    raise err
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_core/shortcuts.py", line 523, in validate_apicall_response
    return v.validate(request, response)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/openapi_core/shortcuts.py", line 446, in validate_response
    validate_apicall_response(
  File "/home/gerald/co/openeo-conformance-test/openeo_conformance_test/conformance_testing.py", line 243, in test_MOCK_processes (Current frame)
    validate_response(
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/python.py", line 194, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/python.py", line 1799, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/runner.py", line 169, in pytest_runtest_call
    item.runtest()
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/runner.py", line 262, in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/runner.py", line 341, in from_call
    result: Optional[TResult] = func()
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/runner.py", line 261, in call_runtest_hook
    return CallInfo.from_call(
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/runner.py", line 222, in call_and_report
    call = call_runtest_hook(item, when, **kwds)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/runner.py", line 133, in runtestprotocol
    reports.append(call_and_report(item, "call", log))
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/runner.py", line 114, in pytest_runtest_protocol
    runtestprotocol(item, nextitem=nextitem)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/main.py", line 348, in pytest_runtestloop
    item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/main.py", line 323, in _main
    config.hook.pytest_runtestloop(session=session)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/main.py", line 269, in wrap_session
    session.exitstatus = doit(config, session) or 0
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/main.py", line 316, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_callers.py", line 80, in _multicall
    res = hook_impl.function(*args)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_manager.py", line 112, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/pluggy/_hooks.py", line 433, in __call__
    return self._hookexec(self.name, self._hookimpls, kwargs, firstresult)
  File "/home/gerald/.cache/pypoetry/virtualenvs/openeo-conformance-test-iNfhDeti-py3.10/lib/python3.10/site-packages/_pytest/config/__init__.py", line 166, in main
    ret: Union[ExitCode, int] = config.hook.pytest_cmdline_main(
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
TypeError: unhashable type: 'dict'

**Comment by GeraldIr** [comment]
I just revisited this problem and it seems to be fixed as of version 0.18.2
